File: Sources/UI/MaxEditSheet.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// Editing a working max appends a record; it never changes a past measurement.
/// Each hand is independent, including whether it has any value of its own yet.
struct MaxEditSheet: View {
    let grip: GripSpec
    var onSaved: (() -> Void)? = nil
    var onClose: () -> Void

    @Environment(TemplateStore.self) private var templates
    @Environment(\.weightUnit) private var weightUnit
    @State private var draft = MaxEditDraft()
    @State private var loaded = false
    @State private var failed = false
    @State private var editingShared = false

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: Metrics.spacing) {
                    gripHeader
                    handValues
                    if let sharedKg { sharedValue(sharedKg) }
                    earlierRecords
                }
                .padding(.horizontal, Metrics.hPadding)
                .padding(.top, 16)
                .padding(.bottom, Metrics.spacing)
                .frame(maxWidth: Metrics.maxContentWidth)
                .frame(maxWidth: .infinity)
            }
            .scrollDismissesKeyboard(.interactively)
            .scrollEdgeEffectStyle(.soft, for: .bottom)
            .background { AppBackground() }
            .navigationTitle("Edit maxes")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { onClose() }
                        .accessibilityIdentifier("maxEdit.cancel")
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Save") { save() }
                        .bold()
                        .disabled(!loaded || !draft.canSave)
                        .accessibilityIdentifier("maxEdit.save")
                }
            }
            .sheet(isPresented: $editingShared) {
                MaxEntrySheet(seed: grip, side: .both) { editingShared = false }
            }
        }
        .onAppear { loadOnce() }
        .onChange(of: templates.maxTable) { _, current in
            guard loaded else { return }
            draft.rebase(leftKg: current.exact(grip: grip.key, side: .left),
                         rightKg: current.exact(grip: grip.key, side: .right))
        }
    }

    private var gripHeader: some View {
        HStack(spacing: 14) {
            FingerGlyph(fingers: grip.fingers, position: grip.position, dot: 10, gap: 4)
                .padding(12)
                .background(Ink.primary.opacity(0.05),
                            in: RoundedRectangle(cornerRadius: Metrics.radiusInner,
                                                 style: .continuous))
            Text(grip.displayName)
                .font(.system(.title3, weight: .semibold))
                .foregroundStyle(Ink.primary)
                .fixedSize(horizontal: false, vertical: true)
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(grip.spoken)
    }

    private var handValues: some View {
        MaterialCard {
            VStack(alignment: .leading, spacing: 16) {
                handValue(.left, kilograms: $draft.leftKg)
                Divider()
                handValue(.right, kilograms: $draft.rightKg)
                Text("Percentage targets follow the values you save. Earlier records stay in your history.")
                    .font(.system(.footnote))
                    .foregroundStyle(Ink.secondary)
                    .fixedSize(horizontal: false, vertical: true)
                if draft.hasInvalidChanges {
                    Text("Enter a max above zero. To remove a record, open Earlier records.")
                        .font(.system(.footnote))
                        .foregroundStyle(Accent.alarm)
                        .fixedSize(horizontal: false, vertical: true)
                }
                if failed {
                    Text("Couldn’t save your maxes. Your changes are still here—please try again.")
                        .font(.system(.footnote))
                        .foregroundStyle(Accent.alarm)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
        }
    }

    private func handValue(_ side: Side, kilograms: Binding<Double>) -> some View {
        ValueRow(
            title: side == .left ? String(localized: "Left hand") : String(localized: "Right hand"),
            unit: weightUnit.symbol,
            value: weightUnit.binding(kilograms),
            range: weightUnit.sliderRangeFromKg(0...100),
            limit: weightUnit.rangeFromKg(0...250),
            step: 0.5,
            decimals: 1,
            caption: draft.originalKg(for: side) == nil && kilograms.wrappedValue == 0
                ? String(localized: "Not set. Enter a value to save a max for this hand.") : nil
        )
    }

    private var sharedKg: Double? {
        templates.currentMaxes[MaxTable.key(grip: grip.key, side: .both)]?.kg
    }

    private func sharedValue(_ kg: Double) -> some View {
        MaterialCard {
            VStack(alignment: .leading, spacing: 8) {
                HStack(alignment: .firstTextBaseline, spacing: 12) {
                    Text("Shared max")
                        .font(.system(.subheadline, weight: .semibold))
                    Spacer(minLength: 8)
                    Text(weightUnit.text(kg))
                        .font(.system(.subheadline, weight: .semibold))
                        .monospacedDigit()
                }
                .foregroundStyle(Ink.primary)
                Text("Used when a hand has no max of its own, and for two-handed pulls.")
                    .font(.system(.footnote))
                    .foregroundStyle(Ink.secondary)
                    .fixedSize(horizontal: false, vertical: true)
                Button {
                    editingShared = true
                } label: {
                    Text("Edit shared max")
                        .font(.system(.subheadline, weight: .semibold))
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .frame(minHeight: 44)
                        .contentShape(.rect)
                }
                .buttonStyle(PressFeedbackButtonStyle())
                .foregroundStyle(Accent.graphite)
                .accessibilityIdentifier("maxEdit.shared")
            }
        }
    }

    private var earlierRecords: some View {
        NavigationLink {
            MaxesView(grip: grip)
        } label: {
            HStack(spacing: 12) {
                Text("Earlier records")
                    .font(.system(.subheadline, weight: .semibold))
                Spacer(minLength: 8)
                Image(systemName: "chevron.right")
                    .font(.system(.footnote, weight: .semibold))
                    .foregroundStyle(Ink.tertiary)
                    .accessibilityHidden(true)
            }
            .foregroundStyle(Accent.graphite)
            .frame(minHeight: 44)
            .contentShape(.rect)
        }
        .buttonStyle(PressFeedbackButtonStyle())
        .accessibilityIdentifier("maxEdit.history")
    }

    private func loadOnce() {
        guard !loaded else { return }
        draft = MaxEditDraft(
            leftKg: templates.currentMaxes[MaxTable.key(grip: grip.key, side: .left)]?.kg,
            rightKg: templates.currentMaxes[MaxTable.key(grip: grip.key, side: .right)]?.kg
        )
        loaded = true
    }

    private func save() {
        guard loaded, draft.canSave else { return }
        let values = draft.changes.map {
            TemplateStore.MaxSave(grip: grip, side: $0.side, kg: $0.kg, source: .manual)
        }
        guard templates.recordMaxes(values) else {
            failed = true
            return
        }
        onSaved?()
        onClose()
    }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/ui/maxes/MaxEditSheet.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.
package run.nuri.getagrip.ui.maxes

import androidx.compose.foundation.focusGroup
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.outlined.RadioButtonUnchecked
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.engine.*
import run.nuri.getagrip.store.LocalHistoryFeed
import run.nuri.getagrip.store.LocalTemplateStore
import run.nuri.getagrip.store.TemplateStore
import run.nuri.getagrip.ui.components.*
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics
import run.nuri.getagrip.ui.theme.readablePageWidth
import run.nuri.getagrip.ui.units.WeightUnits

@Composable
fun MaxEditSheet(grip: GripSpec, onSaved: () -> Unit = {}, onClose: () -> Unit) {
    val templates = LocalTemplateStore.current
    val feed = LocalHistoryFeed.current
    val palette = LocalGripPalette.current
    val scope = rememberCoroutineScope()
    val focus = LocalFocusManager.current
    val haptics = LocalHapticFeedback.current
    var draft by remember(grip.key) { mutableStateOf(MaxEditDraft(
        leftKg = templates.maxTable.exact(grip.key, Side.left),
        rightKg = templates.maxTable.exact(grip.key, Side.right))) }
    var failed by remember { mutableStateOf(false) }
    var committed by remember { mutableStateOf(false) }
    var hasInputFocus by remember { mutableStateOf(false) }
    var history by remember { mutableStateOf(false) }
    var editingShared by remember { mutableStateOf(false) }
    var receipt by remember { mutableStateOf<TemplateStore.MaxSaveReceipt?>(null) }
    val submission = remember { SubmissionState() }
    val current = templates.maxTable
    LaunchedEffect(current, grip.key) {
        draft = draft.copy().also { it.rebase(current.exact(grip.key, Side.left), current.exact(grip.key, Side.right)) }
    }

    fun finish() { onSaved(); onClose() }
    fun save() {
        if (committed || submission.isRunning) return
        submission.launch(scope) {
            focus.clearFocus(force = true)
            withFrameNanos { }
            if (!draft.canSave) return@launch
            failed = false
            val saved = templates.recordMaxesWithReceipt(draft.changes.map {
                TemplateStore.MaxSave(grip, it.side, it.kg, MaxSource.manual)
            })
            if (saved == null) { failed = true; return@launch }
            committed = true
            feed.refresh()
            haptics.performHapticFeedback(HapticFeedbackType.Confirm)
            if (saved.hasDetails) receipt = saved else finish()
        }
    }

    receipt?.let { MaxSaveReceiptScreen(it, onDone = ::finish); return }
    if (editingShared) {
        SharedMaxSheet(grip, onClose = { editingShared = false })
        return
    }
    if (history) {
        MaxesFlowScaffold(tr("Earlier records"), onClose = { history = false }) { padding ->
            CompositionLocalProvider(LocalFloatingTabBarInset provides 0.dp) {
                MaxesListScreen(onAddMax = {}, grip = grip, modifier = Modifier.padding(padding), feed = feed)
            }
        }
        return
    }
    MaxesFlowScaffold(tr("Edit maxes"), onClose = onClose, closeEnabled = !submission.isRunning, actions = {
        TextButton(onClick = ::save, enabled = !committed && !submission.isRunning && (draft.canSave || hasInputFocus),
            modifier = Modifier.testTag("maxEdit.save")) { Text(tr("Save"), fontWeight = FontWeight.SemiBold) }
    }) { padding ->
        Column(Modifier.fillMaxSize().padding(padding).readablePageWidth()
            .verticalScroll(rememberScrollState()).imePadding()
            .onFocusChanged { hasInputFocus = it.hasFocus }.focusGroup()
            .padding(horizontal = Metrics.hPadding, vertical = 16.dp),
            verticalArrangement = Arrangement.spacedBy(Metrics.spacing)) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(14.dp)) {
                GlyphTile(grip, 56.dp)
                Text(grip.displayName, style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold, color = palette.inkPrimary, modifier = Modifier.weight(1f))
            }
            MaxesContentCard {
                listOf(Side.left, Side.right).forEachIndexed { index, side ->
                    if (index > 0) HorizontalDivider()
                    val kg = if (side == Side.left) draft.leftKg else draft.rightKg
                    ValueRow(title = tr(if (side == Side.left) "Left hand" else "Right hand"),
                        value = WeightUnits.fromKg(kg), unit = WeightUnits.symbol,
                        range = WeightUnits.sliderRange(0.0..100.0), limit = WeightUnits.fromKg(0.0..250.0),
                        step = 0.5, decimals = 1, caption = manualBandCaption(kg),
                        modifier = Modifier.testTag("maxEdit.value.${side.rawValue}")) { displayValue ->
                        draft = draft.copy().also {
                            if (side == Side.left) it.leftKg = WeightUnits.toKg(displayValue)
                            else it.rightKg = WeightUnits.toKg(displayValue)
                        }
                    }
                    if (draft.originalKg(side) == kg) {
                        val selected = draft.recordsAnotherTest(side)
                        TextButton(onClick = {
                            draft = draft.copy().also { it.setRecordsAnotherTest(!selected, side) }
                            haptics.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                        }, modifier = Modifier.fillMaxWidth().heightIn(min = 48.dp)
                            .testTag("maxEdit.retest.${side.rawValue}").semantics { this.selected = selected }) {
                            Icon(if (selected) Icons.Filled.CheckCircle else Icons.Outlined.RadioButtonUnchecked, null)
                            Spacer(Modifier.width(8.dp))
                            Text(tr("Record another test"), modifier = Modifier.weight(1f))
                        }
                    }
                }
                if (draft.hasInvalidChanges) Text(tr("Enter a max above zero. To remove a record, open Earlier records."),
                    style = MaterialTheme.typography.bodySmall, color = palette.alarm)
                if (failed) Text(tr("Couldn’t save your maxes. Your changes are still here—please try again."),
                    style = MaterialTheme.typography.bodySmall, color = palette.alarm)
            }
            templates.maxTable.exact(grip.key, Side.both)?.let { shared ->
                MaxesContentCard {
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        Text(tr("Shared max"), fontWeight = FontWeight.SemiBold, modifier = Modifier.weight(1f))
                        Spacer(Modifier.width(12.dp))
                        Text(WeightUnits.text(shared), fontWeight = FontWeight.SemiBold)
                    }
                    Text(tr("Used when a hand has no max of its own, and for two-handed pulls."),
                        style = MaterialTheme.typography.bodySmall, color = palette.inkSecondary)
                    TextButton(onClick = { focus.clearFocus(); editingShared = true }, modifier = Modifier.testTag("maxEdit.shared")) {
                        Text(tr("Edit shared max"))
                    }
                }
            }
            SecondaryButton(tr("Earlier records"), modifier = Modifier.fillMaxWidth().testTag("maxEdit.history")) {
                focus.clearFocus(); history = true
            }
        }
    }
}

internal fun manualBandCaption(kg: Double): String? {
    if (kg <= 0) return L10n.tr("Not set. Enter a value to save a max for this hand.")
    val band = PlanMath.suggestedBand(kg) ?: return null
    return WeightUnits.tr("20–30 %% of that is %s–%s kg", WeightUnits.number(band.start), WeightUnits.number(band.endInclusive))
}
